### ISSUE TITLE
Class typevars lose information from `isinstance` checks

### ISSUE THREAD

**Issue by msullivan** [python/mypy#5720]
```
from typing import Generic, TypeVar

T = TypeVar('T')

class C(Generic[T]):
    def foo(self, x: T) -> None:
        if isinstance(x, int):
            self.bar(x)

    def bar(self, x: T) -> None:
        pass
```
fails with `Argument 1 to "bar" of "C" has incompatible type "int"; expected "T"`, even though `x` is obviously a `T` (that was its annotated type!)

Arguably this is just the Dreaded Intersection Bug (#3603) in a different manifestation?

**Comment by ilevkivskyi** [comment]
Yes, and this is probably a hardest aspect of the problem, because I don't see an easy way to work around this using some fake `TypeInfo`s for intersections.
